File: 198-house-robber/house-robber.java
class Solution {
    int[] dp;
    public int rob(int[] nums) {
        dp = new int[nums.length + 1];
        Arrays.fill(dp , -1);
        return solve(0 , nums);
    }

    private int solve(int i  , int[] nums){
        if(i >= nums.length){
            return 0;
        }

        if(dp[i] != -1){
            return dp[i];
        }

        int pick = nums[i] + solve(i+2 , nums);
        int np = solve(i+1 , nums);
        return dp[i] = Math.max(pick ,np);
    }
}

File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0]*n

        dp[n-1] = nums[n-1]
        dp[n-2] = max(nums[n-2] , nums[n-1])

        if n == 2:
            return max(dp[n-1] , dp[n-2])

        for i in range(n-3 , -1 , -1):
            take = nums[i] + dp[i+2]
            not_take = dp[i+1]
            dp[i] = max(take , not_take)
        
        return dp[0]